source/docs/software/advanced-gradlerio/profiling-with-visualvm.rst

```diff
@@ -0,0 +1,168 @@
+
+Setting up Gradle
+-----------------
+
+GradleRIO supports passing JVM launch arguments, and this is what is necessary to enable remote debugging. Remote debugging is a feature that allows a local machine (such as the user's desktop) to view important information about a remote target (in our case, a roboRIO). To begin, locate the ``frcJava`` code block located in the projects ``build.gradle``. Below is what is looks like.
+
+.. code-block:: groovy
+
+   deploy {
+      targets {
+         roborio(getTargetTypeClass('RoboRIO')) {
+               // Team number is loaded either from the .wpilib/wpilib_preferences.json
+               // or from command line. If not found an exception will be thrown.
+               // You can use getTeamOrDefault(team) instead of getTeamNumber if you
+               // want to store a team number in this file.
+               team = project.frc.getTeamNumber()
+               debug = project.frc.getDebugOrDefault(false)
+
+               artifacts {
+                  // First part is artifact name, 2nd is artifact type
+                  // getTargetTypeClass is a shortcut to get the class type using a string
+
+                  frcJava(getArtifactTypeClass('FRCJavaArtifact')) {
+                  }
+
+                  // Static files artifact
+                  frcStaticFileDeploy(getArtifactTypeClass('FileTreeArtifact')) {
+                     files = project.fileTree('src/main/deploy')
+                     directory = '/home/lvuser/deploy'
+                  }
+               }
+         }
+      }
+   }
+
+We will be replacing
+
+.. code-block:: groovy
+
+               frcJava(getArtifactTypeClass('FRCJavaArtifact')) {
+               }
+
+with
+
+.. code-block:: groovy
+
+   frcJava(getArtifactTypeClass('FRCJavaArtifact')) {
+      // Enable VisualVM connection
+      jvmArgs.add("-Dcom.sun.management.jmxremote=true")
+      jvmArgs.add("-Dcom.sun.management.jmxremote.port=1198")
+      jvmArgs.add("-Dcom.sun.management.jmxremote.local.only=false")
+      jvmArgs.add("-Dcom.sun.management.jmxremote.ssl=false")
+      jvmArgs.add("-Dcom.sun.management.jmxremote.authenticate=false")
+      jvmArgs.add("-Djava.rmi.server.hostname=10.XX.XX.2") // Replace XX.XX with team number
+   }
+
+We are adding a few arguments here. In order:
+
+* Enable remote debugging
+* Set the remote debugging port to 1198
+* Allow listening from remote targets
+* Disable SSL authentication being required
+* Set the hostname to the roboRIOs team number. Be sure to replace this.
+
+.. important:: The hostname when connected via USB-B should be ``172.22.11.2``.
+
+Running VisualVM
+----------------
+
+Launching VisualVM is done via the commandline with a few parameters. First, we navigate to the directory containing VisualVM. Then, launch it with parameters, passing it the WPILib JDK path. On a Windows machine, it looks like the following:
```

Comment:
ditto

Reply:
Huh?

Reply:
ditto meaning "Add options for different OSs via a switcher"